lib/wanda/execution/evaluation.ex

```diff
@@ -0,0 +1,169 @@
+    agents_results =
+      Enum.map(agents_facts, fn {agent_id, facts} ->
+        %AgentCheckResult{agent_id: agent_id, facts: facts}
+        |> add_agent_expectation_result(check_id)
+      end)
+
+    %CheckResult{check_result | agents_check_results: agents_results}
+  end
+
+  defp add_agent_expectation_result(
+         %AgentCheckResult{facts: facts} = agent_check_result,
+         check_id
+       ) do
+    expectation_evaluations =
+      check_id
+      |> Catalog.get_expectations()
+      |> Enum.map(&eval_expectation(&1, facts))
+
+    %AgentCheckResult{
+      agent_check_result
+      | expectation_evaluations: expectation_evaluations
+    }
+  end
+
+  defp eval_expectation(%Expectation{name: name, type: type, expression: expression}, facts) do
+    case Abacus.eval(expression, facts) do
+      {:ok, return_value} ->
+        %ExpectationEvaluation{name: name, type: type, return_value: return_value}
+
+      {:error, :einkey} ->
+        %ExpectationEvaluationError{
+          name: name,
+          message: "Fact is not present.",
+          type: :fact_missing_error
+        }
+
+      _ ->
+        %ExpectationEvaluationError{
+          name: name,
+          message: "Illegal expression provided, check expression syntax.",
+          type: :illegal_expression_error
+        }
+    end
+  end
+
+  defp add_expectation_evaluations(
+         %CheckResult{check_id: check_id, agents_check_results: agents_check_results} = result
+       ) do
+    expectation_results =
+      agents_check_results
+      |> Enum.flat_map(fn %AgentCheckResult{expectation_evaluations: expectation_evaluations} ->
+        expectation_evaluations
+      end)
+      |> Enum.group_by(& &1.name)
+      |> Enum.map(fn {name, expectation_evaluations} ->
+        %{type: type} = Catalog.get_expectation(check_id, name)
+
+        %ExpectationResult{
+          name: name,
+          type: type,
+          result: eval_expectation_result_or_error(type, expectation_evaluations)
+        }
+      end)
+
+    %CheckResult{result | expectation_results: expectation_results}
+  end
+
+  defp eval_expectation_result_or_error(type, expectation_evaluations) do
+    if has_error?(expectation_evaluations) do
+      false
+    else
+      eval_expectation_result(type, expectation_evaluations)
+    end
+  end
+
+  defp has_error?(expectation_evaluations) do
+    Enum.any?(expectation_evaluations, fn
+      %ExpectationEvaluationError{} -> true
+      _ -> false
+    end)
+  end
+
+  defp eval_expectation_result(:expect_same, expectation_evaluations) do
+    expectation_evaluations
+    |> Enum.uniq_by(& &1.return_value)
+    |> Kernel.length() == 1
+  end
+
+  defp eval_expectation_result(:expect, expectations_evaluations) do
+    Enum.all?(expectations_evaluations, &(&1.return_value == true))
+  end
+
+  defp aggregate_result(%CheckResult{expectation_results: expectation_results} = check_result) do
```

Comment:
Maybe better `aggregate_check_result` ? To clearly identify this function as something that works on an individual check

Reply:
thanks, makes sense